User: fetch only on app boot

jsnajdr opened the pull request:
This patch removes automatic call to `user.fetch()` every time the `user.get()` getter is called and the user data are not available (i.e., not logged in or not initialized yet).

Instead, we call `user.fetch()` only when calling `user.initialize()` at Calypso boot and at a few places that call it explicitly to reload the updated user object.

This change should make problems like these in #36519 much more tractable -- no more uncontrolled user fetching at random places, especially in Signup.

I also removed code that stores the user object to local storage key `wpcom_user` and tries to load it from there on boot. That code is quite useless and only causes trouble: I experienced several times a weird Calypso state where the user is logged out (i.e., doesn't have session cookies), but valid user is loaded from `wpcom_user`, making Calypso think it's logged in.

The only supported ways how to get the user info is server-side bootstrap and the `/me` endpoint.

Now we only store `wpcom_user_id` in local storage and the only purpose is to clear local storage when a different user logs in.


**How to test:**
Test that the support sessions continue to work -- need help from @brandonpayton 🙂 

Test that Signup continues to work, especially the logged-out -> logged-in transitions. Something for @andrewserong and @Automattic/zelda 

Another place where I had to work hard is accepting invites by user who doesn't have an account yet and is signing up at the same time. e2e tests were failing and it took me several hours to figure out that I need to change `!==` to `!=` in 27f94b2 🙂

jsnajdr committed:
Fixup tests of user/utils: always mock user, remove chai and sinon

jsnajdr committed:
User: remove legacy code for dealing with settings

jsnajdr committed:
CommunityTranslator: stop listening for User changes

jsnajdr committed:
TranslatorJumpstart: stop listening to User changes

jsnajdr committed:
Stop loading from local storage (wpcom_user), store only ID to detect clearing

jsnajdr committed:
User: fetch only on initialization and explicit calls

And stop fetching on every `user.get()` call.

jsnajdr committed:
User: don't clear storage when previously stored ID is undefined

Checking for `=== null` was wrong: if the key doesn't exist, `store.get` returns `undefined`,
unlike the native `localStorage.getItem` that returns `null`.

The incorrect check caused local storage to be cleared after logging in. When accepting an
invite in logged-out session, that means that the `invite_accepted` local storage key was
deleted after signup, causing the process to fail.

This should fix failing e2e tests for invites.

sgomes commented on client/lib/user/user.js:
```diff
@@ -107,9 +93,9 @@ User.prototype.initialize = function() {
+	const storedUserId = store.get( 'wpcom_user_id' );
 
-	if ( storedUser && storedUser.ID !== userId ) {
+	if ( storedUserId != null && storedUserId !== userId ) {
```
I suggest making this a bit more explicit; you're relying on behaviour that isn't immediately obvious at first glance. This code is vulnerable to well-intentioned drive-by fixes that want to remove `!=` from the codebase because of lint checks 🙂 

How about:
```suggestion
	if ( storedUserId !== null && storedUserId !== undefined && storedUserId !== userId ) {
```

jsnajdr replied:
`!= null` is a valid use of the `!=` operator, the [eqeqeq](https://eslint.org/docs/rules/eqeqeq) rule has an `allow-null` exception for it, and our ESLint config has it enabled. I recently learned about it from @ljharb in #36256 🙂

sgomes reviewed (commented):
These changes make sense and simplify things significantly. Thanks for working on them, @jsnajdr!
They LGTM, but I'll leave approving to the folks performing the expert testing.

I left one minor nitpick type comment.

brandonpayton reviewed (changes requested):
Calypso appears to boot just fine for support sessions, with and without user bootstrapping. Overall, this looks like a solid improvement!

It looks like one component will be broken by this change, and I added a note for that.

jsnajdr committed:
user.fetch doesn't need to be async

jsnajdr commented on client/lib/user/user.js:
```diff
@@ -133,26 +116,34 @@ User.prototype.get = function() {
  * @uses `wpcom`
  */
-User.prototype.fetch = function() {
+User.prototype.fetch = async function() {
```
Good observation. It's a sync function that returns a promise. I probably originally planned to do some `await` in there, but then I didn't and forgot to remove the `async` keyword. Fixed.

jsnajdr committed:
RedirectWhenLoggedIn: fix detecting login in another tab

The `wpcom_user` local storage key with the whole user object is gone. Rewrite
the login-detecting code to simply check that a login happened, without checking
the user ID or email.

jsnajdr commented on client/lib/user/user.js:
```diff
@@ -185,12 +171,11 @@ User.prototype.handleFetchFailure = function( error ) {
-	// Store user info in `this.data` and localstorage as `wpcom_user`
-	store.set( 'wpcom_user', userData );
+	// Store user ID in local storage so that we can detect a change and clear the storage
+	store.set( 'wpcom_user_id', userData.ID );
```
Good catch! I'm fixing that in 68f73ec, where I check the new `wpcom_user_id` key instead of the old `wpcom_user` that's gone.

One downside is that the `RedirectWhenLoggedIn` component now doesn't verify that the user with the expected ID and email has logged in, but will accept **any** login. I don't think that's a huge concern though. Implementing the check without the email being stored in local storage would be very difficult, as Calypso can't switch from logged-out to logged-in state without a full reload, and cannot retrieve the user info from the `/me` endpoint without being logged-in from the very beginning. Cc @jblz 

I tested this patch by requesting a magic login link, then logging in in second tab, and verifying that the first tab detects the login and automatically reloads and navigates to `/help`.

jsnajdr commented:
@brandonpayton All issues fixed, ready for another review round.

brandonpayton reviewed (approved):
The RedirectWhenLoggedIn fix looks good, and support sessions continue to test well.

andrewserong reviewed (approved):
Thanks so much for this @jsnajdr — I've tested across creating new accounts with the normal user step, with the passwordless signup A/B test, creating various new sites, logging in and logging out, with redirects, and connecting a Jetpack site, and can't find any faults or edge cases in signup.

This is looking great to me 👍

creativecoder commented:
Thank you for fixing this! I've run into issues that are likely related to user.fetch more than once.
